rfcs/3-add-event-notification-flags.md

```diff
@@ -0,0 +1,232 @@
+in the original Redis enumeration variants, which we shouldn't expose
+to the user.
+
+Another drawback is that JavaScript doesn't have any means to check the
+values for incompatibility with each other or any other sort of overlaps.
+This will have to be done at deploy time, by RedisGears itself.
+
+## Option 2 (Expose the event command)
+
+However, if we already think about the first approach, we might as well
+reconsider the event command argument of the notification, and expose it
+too; just as well as we could allow the user's key space trigger to
+subscribe only to the commands wanted when those were fired.
+
+This also requires bumping the minor version of the JS API, and the
+changes can be performed in a backwards-compatible way.
+
+To implement this solution, we need to expose an optional argument which
+may consist of the command name(s) as strings. Once there is an
+event notification, the RedisGears will look for a key space trigger
+registered with the event notification command names, filter the ones
+which have subscribed to this particular command event, and only invoke
+these.
+
+Example:
+
+```JS
+#!js name=lib api_version=1.2
+
+redis.registerKeySpaceTrigger(
+  'onlyzsetnotifications', // trigger name
+  '', //key prefix
+  function(client, data) {
+    console.log("Got this key data created1: " + data);
+  }, //callback
+  {
+    description: 'description',
+    onTriggerFired: function(client, data) {
+        console.log("Got this key data created2: " + data);
+    },
+    // The new optional argument to specify the event notification
+    // command names:
+    eventNotificationCommands: ["ZSET"]
+  } //optional arguments
+)
+```
+
+Maintaining a list of all possible Redis commands is unnecessary: even
+though possible, it brings a burden to maintain while not bringing any
+major user experience improvement.
+
+
+## Option 3 (1 & 2 combined)
+
+The third option is just to combine both the implementations into one.
+
+We can change the `redis.registerKeySpaceTrigger` function to accept
+both, the flags and the command names.
+
+Example:
+
+
+```JS
+#!js name=lib api_version=1.2
+
+redis.registerKeySpaceTrigger(
+  'onlynewzsetnotifications', // trigger name
+  '', //key prefix
+  function(client, data) {
+    console.log("Got this key data created1: " + data);
+  }, //callback
+  {
+    description: 'description',
+    onTriggerFired: function(client, data) {
+        console.log("Got this key data created2: " + data);
+    },
+    // The new optional argument to specify the event notification flags:
+    eventNotificationFlags: [EventNotificationFlags.NEW],
+    // The new optional argument to specify the event notification
+    // command names:
+    eventNotificationCommands: ["ZSET"]
+  } //optional arguments
+)
+```
+
+In this case, however, there is again, might be a possible overlap of
+the values of both, that needs to be checked at deploy time.
+
+## Option 4 (Listen to the "NEW" flag by default)
```

Comment:
Other than being a breaking change, did you consider the other pros and cons of option 4 vs the others? I guess having a single function to handle all cases can be an advantage. But could that lead to performance issues, if, for example, the function doesn't care about some events?

Reply:
So right now, we don't have any filtering w.r.t. the flags, and we only filter by key name or key prefix while iterating over an array of user callbacks. There is no order within this array, so nothing is done to make the search faster. If we go this way, perhaps it makes sense to store a hashmap where the keys would be compatible event flags/event command names, and the keys would be a list of these user callbacks so we can look up the needed callbacks faster rather than going through all the callbacks iteratively. Perhaps the same optimisation could be done for the key names/prefixes; I need to check.

To summarise and answer the last question, if a function doesn't care about some events, we still can have a good performance optimisation to look up such callbacks and fire them quickly. However, the implicated cost of the finding itself, navigating to the callbacks objects, firing one by one, all the indirections, will still (and always) cause higher latency, for there still some processing takes place, and it is not for free.

What do you think? @MeirShpilraien